tests/components/sma/test_config_flow.py

```diff
@@ -59,90 +92,103 @@ async def test_form_cannot_connect(hass: HomeAssistant) -> None:
+        patch("pysma.SMA.device_info", return_value=MOCK_DEVICE),
         _patch_async_setup_entry() as mock_setup_entry,
     ):
         result = await hass.config_entries.flow.async_configure(
             result["flow_id"],
-            MOCK_USER_INPUT,
+            MOCK_DHCP_DISCOVERY_INPUT,
         )
+        await hass.async_block_till_done()
 
-    assert result["type"] is FlowResultType.FORM
-    assert result["errors"] == {"base": "cannot_retrieve_device_info"}
-    assert len(mock_setup_entry.mock_calls) == 0
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == MOCK_DHCP_DISCOVERY["host"]
+    assert result["data"] == MOCK_DHCP_DISCOVERY
+
+    assert len(mock_setup_entry.mock_calls) == 1
 
 
-async def test_form_unexpected_exception(hass: HomeAssistant) -> None:
-    """Test we handle unexpected exception."""
+async def test_dhcp_already_configured(
+    hass: HomeAssistant, mock_config_entry: MockConfigEntry
+) -> None:
+    """Test starting a flow by dhcp when already configured."""
     result = await hass.config_entries.flow.async_init(
-        DOMAIN, context={"source": SOURCE_USER}
+        DOMAIN, context={"source": SOURCE_DHCP}, data=DHCP_DISCOVERY_DUPLICATE
     )
 
-    with (
-        patch("pysma.SMA.new_session", side_effect=Exception),
-        _patch_async_setup_entry() as mock_setup_entry,
-    ):
-        result = await hass.config_entries.flow.async_configure(
-            result["flow_id"],
-            MOCK_USER_INPUT,
-        )
-
-    assert result["type"] is FlowResultType.FORM
-    assert result["errors"] == {"base": "unknown"}
-    assert len(mock_setup_entry.mock_calls) == 0
-
+    assert result["type"] is FlowResultType.ABORT
+    assert result["reason"] == "already_configured"
 
-async def test_form_already_configured(hass: HomeAssistant, mock_config_entry) -> None:
-    """Test starting a flow by user when already configured."""
-    mock_config_entry.add_to_hass(hass)
 
+@pytest.mark.parametrize(
+    ("exception", "error"),
+    [
+        (SmaConnectionException, "cannot_connect"),
+        (SmaAuthenticationException, "invalid_auth"),
+        (SmaReadException, "cannot_retrieve_device_info"),
+        (Exception, "unknown"),
+    ],
+)
+async def test_dhcp_exceptions(
+    hass: HomeAssistant, exception: Exception, error: str
+) -> None:
+    """Test we handle cannot connect error."""
     result = await hass.config_entries.flow.async_init(
-        DOMAIN, context={"source": SOURCE_USER}
+        DOMAIN,
+        context={"source": SOURCE_DHCP},
+        data=DHCP_DISCOVERY,
     )
 
     with (
-        patch("pysma.SMA.new_session", return_value=True),
-        patch("pysma.SMA.device_info", return_value=MOCK_DEVICE),
-        patch("pysma.SMA.close_session", return_value=True),
+        patch("pysma.SMA.new_session", side_effect=exception),
         _patch_async_setup_entry() as mock_setup_entry,
```

Comment:
the mock setup entry is not needed here, as we don't run that here (as it fails)